File: Python/programmers/etc/level3/퍼즐_조각_채우기.py
from copy import deepcopy

dx, dy = [-1, 0, 1, 0], [0, 1, 0, -1]

def dfs(graph, x, y, pos, n, num):
    ret = [pos]
    for d in range(4):
        nx, ny = x+dx[d], y+dy[d]
        if (0 <= nx < n) and (0 <= ny < n) and graph[nx][ny] == num:
            graph[nx][ny] = 2
            ret += dfs(graph, nx, ny, [pos[0]+dx[d], pos[1]+dy[d]], n, num)
    return ret

def rotate(arr):
    n = len(arr)
    ret = [[0]*n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            ret[j][n-1-i] = arr[i][j]
    return ret

def solution(game_board, table):
    answer = 0
    board_copy = deepcopy(game_board)

    n = len(game_board)
    block = []

    for i in range(n):
        for j in range(n):
            if board_copy[i][j] == 0:
                board_copy[i][j] = 2
                res = dfs(board_copy, i, j, [0, 0], n, 0)[1:]
                block.append(res)
    for _ in range(4):
        table = rotate(table)
        table_copy = deepcopy(table)
        for i in range(n):
            for j in range(n):
                if table_copy[i][j] == 1:
                    table_copy[i][j] = 2
                    res = dfs(table_copy, i, j, [0, 0], n, 1)[1:]
                    if res in block:
                        block.remove(res)
                        answer += len(res)+1
                        table = deepcopy(table_copy)
                    else:
                        table_copy = deepcopy(table)
    return answer


game_board = [[1,1,0,0,1,0],[0,0,1,0,1,0],[0,1,1,0,0,1],[1,1,0,1,1,1],[1,0,0,0,1,0],[0,1,1,1,0,0]]
table = [[1,0,0,1,1,0],[1,0,1,0,1,0],[0,1,1,0,1,1],[0,0,1,0,0,0],[1,1,0,1,1,0],[0,1,0,0,0,0]]
print(solution(game_board, table)) # 14

game_board = [[0,0,0],[1,1,0],[1,1,1]]
table = [[1,1,1],[1,0,0],[0,0,0]]
print(solution(game_board, table)) # 0

File: Java/programmers/etc/level3/퍼즐조각채우기.java
package CodeTest.Java.programmers.etc.level3;

import java.util.*;

class Point implements Comparable<Point> {
    int x, y;

    Point(int x, int y) {
        this.x = x;
        this.y = y;
    }

    @Override
    public int compareTo(Point o) {
        return this.x == o.x ? this.y - o.y : this.x - o.x; 
    }
}

public class 퍼즐조각채우기 {
    private static int[] dx = {-1,0,1,0};
    private static int[] dy = {0,-1,0,1};

    private static int n;
    private static List<List<Point>> empty = new ArrayList<>();
    private static List<List<Point>> block = new ArrayList<>();
    private static boolean[][] visited;

    public int solution(int[][] game_board, int[][] table) {
        n = game_board.length;
        visited = new boolean[n][n];

        for (int i=0; i<n; i++) {
            for (int j=0; j<n; j++) {
                if (game_board[i][j] == 0 && !visited[i][j]) {
                    empty.add(bfs(game_board, i, j, 0));
                }
            }
        }

        for (int i=0; i<n; i++) Arrays.fill(visited[i], false);

        for (int i=0; i<n; i++) {
            for (int j=0; j<n; j++) {
                if (table[i][j] == 1 && !visited[i][j]) {
                    block.add(bfs(table, i, j, 1));
                }
            }
        }

        int answer = 0;
        boolean[] check = new boolean[empty.size()];

        for (int i=0; i<block.size(); i++) {
            List<Point> b = block.get(i);
            for (int j=0; j<empty.size(); j++) {
                List<Point> e = empty.get(j);
                if (e.size() == b.size() && !check[j]) {
                    if (rotate(e, b)) {
                        answer += b.size();
                        check[j] = true;
                        break;
                    }
                }
            }
        }

        return answer;
    }

    private List<Point> bfs(int[][] board, int x, int y, int type) {
        visited[x][y] = true;

        Queue<Point> q = new LinkedList<>();
        q.add(new Point(x, y));

        List<Point> result = new ArrayList<>();
        result.add(new Point(0, 0));

        while (!q.isEmpty()) {
            Point cur = q.poll();

            for (int i=0; i<4; i++) {
                int nx = cur.x+dx[i], ny = cur.y+dy[i];
                if (0 <= nx && nx < n && 0 <= ny && ny < n) {
                    if (!visited[nx][ny] && board[nx][ny] == type) {
                        q.add(new Point(nx, ny));
                        visited[nx][ny] = true;
                        result.add(new Point(nx-x, ny-y));
                    }
                }
            }
        }

        Collections.sort(result);

        return result;
    }

    private static boolean rotate(List<Point> empty, List<Point> block) {
        for (int i=0; i<4; i++) {
            int zeroX = block.get(0).x, zeroY = block.get(0).y;
            
            for (int j=0; j<block.size(); j++) {
                block.get(j).x -= zeroX;
                block.get(j).y -= zeroY;
            }

            boolean flag = true;

            for (int j=0; j<empty.size(); j++) {
                Point eP = empty.get(j);
                Point bP = block.get(j);
                if (eP.x != bP.x || eP.y != bP.y) {
                    flag = false;
                    break;
                }
            }

            if (flag) return true;
            else {
                for (int j=0; j<block.size(); j++) {
                    int tmp = block.get(j).x;
                    block.get(j).x = block.get(j).y;
                    block.get(j).y = -tmp;
                }
                Collections.sort(block);
            }
        }

        return false;
    }    
}
